fri/src/verifier.rs

```diff
@@ -129,35 +128,70 @@ where
     let mut folded_eval = F::ZERO;
     let mut ro_iter = reduced_openings.into_iter().peekable();
 
-    for (log_folded_height, (&beta, comm, opening)) in izip!((0..log_max_height).rev(), steps) {
-        if let Some((_, ro)) = ro_iter.next_if(|(lh, _)| *lh == log_folded_height + 1) {
+    let mut log_folded_height = log_max_height;
+
+    while log_folded_height > config.log_blowup + config.log_final_poly_len {
+        if let Some((_, ro)) = ro_iter.next_if(|(lh, _)| *lh == log_folded_height) {
             folded_eval += ro;
         }
 
-        let index_sibling = index ^ 1;
-        let index_pair = index >> 1;
+        let (&beta, comm, opening) = steps.next().unwrap();
+
+        let index_row = index >> config.arity_bits;
+
+        // Verify that `folded_eval` and evals from reduced_openings match opened_rows
+        assert_eq!(folded_eval, opening.opened_rows[0][index % config.arity()]);
```

Comment:
is this a sanity check, or a step in verification? if it's a step in verification, you should return `FriError`